services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusService.kt

```diff
@@ -911,10 +918,67 @@ class HakemusService(
         if (updatedApplicationEntity.hanke.generated) {
             updatedApplicationEntity.hanke.nimi = request.name
         }
+        updateTormaystarkastelut(updatedApplicationEntity)
         sendApplicationNotifications(updatedApplicationEntity, originalContactUserIds, userId)
         return hakemusRepository.save(updatedApplicationEntity)
     }
 
+    /** Calculate the traffic nuisance indexes for each work area of an application. */
+    private fun updateTormaystarkastelut(hakemusEntity: HakemusEntity) {
+        if (hakemusEntity.hakemusEntityData is KaivuilmoitusEntityData) {
+            val hakemusEntityData = hakemusEntity.hakemusEntityData as KaivuilmoitusEntityData
+            if (hakemusEntityData.startTime != null && hakemusEntityData.endTime != null) {
+                hakemusEntity.hakemusEntityData =
+                    hakemusEntityData.copy(
+                        areas =
+                            hakemusEntityData.areas?.map { area ->
+                                updateTormaystarkastelutForArea(
+                                    area,
+                                    hakemusEntityData.startTime.toLocalDate(),
+                                    hakemusEntityData.endTime.toLocalDate(),
+                                )
+                            },
+                    )
+            }
+        }
```

Comment:
Nested blobs are hard to read. This can be simplified greatly by inverting the `if`s to guard clauses and using one or two variables for the innermost values.

You could also get fancy and combine the type check and cast in a when-statement:
```kotlin
        val hakemusEntityData =
            when (val data = hakemusEntity.hakemusEntityData) {
                is KaivuilmoitusEntityData -> data
                is JohtoselvityshakemusEntityData -> return
            }
```

Reply:
Refactored the code a bit. Hopefully it is now more readable.